File: src/robot_fastlio_mapping/src/fastlio_odom_bridge_node.cpp
#include <cmath>
#include <cstdint>
#include <memory>
#include <optional>
#include <string>
#include <utility>

#include "builtin_interfaces/msg/time.hpp"
#include "geometry_msgs/msg/quaternion.hpp"
#include "geometry_msgs/msg/transform_stamped.hpp"
#include "nav_msgs/msg/odometry.hpp"
#include "rclcpp/rclcpp.hpp"
#include "tf2/LinearMath/Quaternion.h"
#include "tf2/LinearMath/Transform.h"
#include "tf2/exceptions.h"
#include "tf2/time.h"
#include "tf2_msgs/msg/tf_message.hpp"
#include "tf2_ros/buffer.h"
#include "tf2_ros/transform_listener.h"

namespace
{
rclcpp::QoS make_qos(const std::size_t depth, const bool reliable)
{
  auto qos = rclcpp::QoS(rclcpp::KeepLast(depth));
  qos.durability_volatile();
  if (reliable) {
    qos.reliable();
  } else {
    qos.best_effort();
  }
  return qos;
}

double normalize_angle(const double value)
{
  return std::atan2(std::sin(value), std::cos(value));
}

tf2::Quaternion quaternion_from_msg(const geometry_msgs::msg::Quaternion & msg)
{
  tf2::Quaternion q(msg.x, msg.y, msg.z, msg.w);
  if (q.length2() <= 1.0e-24) {
    q.setValue(0.0, 0.0, 0.0, 1.0);
  } else {
    q.normalize();
  }
  return q;
}

geometry_msgs::msg::Quaternion quaternion_from_yaw(const double yaw)
{
  tf2::Quaternion q;
  q.setRPY(0.0, 0.0, yaw);
  geometry_msgs::msg::Quaternion msg;
  msg.x = q.x();
  msg.y = q.y();
  msg.z = q.z();
  msg.w = q.w();
  return msg;
}

tf2::Transform transform_from_pose(
  const geometry_msgs::msg::Point & position,
  const geometry_msgs::msg::Quaternion & orientation)
{
  return tf2::Transform(
    quaternion_from_msg(orientation),
    tf2::Vector3(position.x, position.y, position.z));
}

tf2::Transform transform_from_msg(const geometry_msgs::msg::TransformStamped & msg)
{
  const auto & t = msg.transform.translation;
  return tf2::Transform(
    quaternion_from_msg(msg.transform.rotation),
    tf2::Vector3(t.x, t.y, t.z));
}

double yaw_from_transform(const tf2::Transform & transform)
{
  double roll = 0.0;
  double pitch = 0.0;
  double yaw = 0.0;
  transform.getBasis().getRPY(roll, pitch, yaw);
  return yaw;
}

builtin_interfaces::msg::Time time_to_msg(const rclcpp::Time & stamp)
{
  std::int64_t total_ns = stamp.nanoseconds();
  if (total_ns < 0) {
    total_ns = 0;
  }
  builtin_interfaces::msg::Time msg;
  msg.sec = static_cast<std::int32_t>(total_ns / 1000000000LL);
  msg.nanosec = static_cast<std::uint32_t>(total_ns % 1000000000LL);
  return msg;
}
}  // namespace

class FastlioOdomBridge : public rclcpp::Node
{
public:
  FastlioOdomBridge()
  : Node("fastlio_odom_bridge")
  {
    input_topic_ = declare_parameter<std::string>("input_topic", "/Odometry");
    output_topic_ = declare_parameter<std::string>("output_topic", "/fastlio/base_odometry");
    tf_topic_ = declare_parameter<std::string>("tf_topic", "/tf_slam2d");
    output_odom_frame_ = declare_parameter<std::string>("output_odom_frame", "odom");
    output_base_frame_ = declare_parameter<std::string>("output_base_frame", "base_link");
    sensor_frame_ = declare_parameter<std::string>("sensor_frame", "lidar_link");
    anchor_on_first_sample_ = declare_parameter<bool>("anchor_on_first_sample", true);
    flatten_to_2d_ = declare_parameter<bool>("flatten_to_2d", true);
    publish_tf_ = declare_parameter<bool>("publish_tf", false);
    restamp_output_to_now_ = declare_parameter<bool>("restamp_output_to_now", false);
    output_stamp_offset_sec_ = declare_parameter<double>("output_stamp_offset_sec", 0.0);
    input_reliable_ = declare_parameter<bool>("input_reliable", false);
    output_reliable_ = declare_parameter<bool>("output_reliable", true);
    input_qos_depth_ =
      static_cast<int>(std::max<std::int64_t>(1, declare_parameter<std::int64_t>("input_qos_depth", 1)));
    output_qos_depth_ =
      static_cast<int>(std::max<std::int64_t>(1, declare_parameter<std::int64_t>("output_qos_depth", 20)));
    tf_qos_depth_ =
      static_cast<int>(std::max<std::int64_t>(1, declare_parameter<std::int64_t>("tf_qos_depth", 100)));
    static_lookup_timeout_sec_ =
      std::max(0.0, declare_parameter<double>("static_lookup_timeout_sec", 0.05));

    tf_buffer_ = std::make_unique<tf2_ros::Buffer>(get_clock());
    tf_listener_ = std::make_unique<tf2_ros::TransformListener>(*tf_buffer_);

    odom_pub_ = create_publisher<nav_msgs::msg::Odometry>(
      output_topic_, make_qos(static_cast<std::size_t>(output_qos_depth_), output_reliable_));
    if (publish_tf_) {
      tf_pub_ = create_publisher<tf2_msgs::msg::TFMessage>(
        tf_topic_, make_qos(static_cast<std::size_t>(tf_qos_depth_), true));
    }
    odom_sub_ = create_subscription<nav_msgs::msg::Odometry>(
      input_topic_,
      make_qos(static_cast<std::size_t>(input_qos_depth_), input_reliable_),
      std::bind(&FastlioOdomBridge::on_odom, this, std::placeholders::_1));

    RCLCPP_INFO(
      get_logger(),
      "FAST-LIO C++ odom bridge: %s -> %s, frame=%s->%s, input_reliable=%s, "
      "output_reliable=%s, publish_tf=%s, restamp_output_to_now=%s",
      input_topic_.c_str(),
      output_topic_.c_str(),
      output_odom_frame_.c_str(),
      output_base_frame_.c_str(),
      input_reliable_ ? "true" : "false",
      output_reliable_ ? "true" : "false",
      publish_tf_ ? "true" : "false",
      restamp_output_to_now_ ? "true" : "false");
  }

private:
  std::optional<tf2::Transform> lookup_base_from_child(const std::string & child_frame)
  {
    if (child_frame == output_base_frame_) {
      return tf2::Transform(
        tf2::Quaternion(0.0, 0.0, 0.0, 1.0),
        tf2::Vector3(0.0, 0.0, 0.0));
    }
    if (base_from_child_) {
      return base_from_child_;
    }

    try {
      const auto tf = tf_buffer_->lookupTransform(
        output_base_frame_,
        child_frame,
        tf2::TimePointZero,
        tf2::durationFromSec(static_lookup_timeout_sec_));
      base_from_child_ = transform_from_msg(tf);
      RCLCPP_INFO(
        get_logger(),
        "using static %s->%s to convert FAST-LIO odom",
        output_base_frame_.c_str(),
        child_frame.c_str());
      return base_from_child_;
    } catch (const tf2::TransformException & exc) {
      RCLCPP_WARN_THROTTLE(
        get_logger(),
        *get_clock(),
        2000,
        "waiting for static %s->%s: %s",
        output_base_frame_.c_str(),
        child_frame.c_str(),
        exc.what());
      return std::nullopt;
    }
  }

  builtin_interfaces::msg::Time output_stamp(const nav_msgs::msg::Odometry & msg)
  {
    if (!restamp_output_to_now_) {
      return msg.header.stamp;
    }
    auto stamp = get_clock()->now();
    if (std::abs(output_stamp_offset_sec_) > 1.0e-9) {
      stamp = stamp + rclcpp::Duration::from_seconds(output_stamp_offset_sec_);
    }
    return time_to_msg(stamp);
  }

  void on_odom(const nav_msgs::msg::Odometry::SharedPtr msg)
  {
    std::string child_frame = msg->child_frame_id.empty() ? sensor_frame_ : msg->child_frame_id;
    if (!child_frame.empty() && child_frame.front() == '/') {
      child_frame.erase(0, 1);
    }

    const auto base_from_child = lookup_base_from_child(child_frame);
    if (!base_from_child) {
      return;
    }

    const auto odom_from_child =
      transform_from_pose(msg->pose.pose.position, msg->pose.pose.orientation);
    const auto odom_from_base = odom_from_child * base_from_child->inverse();
    auto translation = odom_from_base.getOrigin();
    double x = translation.x();
    double y = translation.y();
    double z = translation.z();
    double yaw = yaw_from_transform(odom_from_base);

    if (anchor_on_first_sample_) {
      if (!anchor_) {
        anchor_ = std::make_tuple(x, y, yaw);
      }
      const auto [anchor_x, anchor_y, anchor_yaw] = *anchor_;
      const double dx = x - anchor_x;
      const double dy = y - anchor_y;
      const double c = std::cos(anchor_yaw);
      const double s = std::sin(anchor_yaw);
      x = c * dx + s * dy;
      y = -s * dx + c * dy;
      yaw = normalize_angle(yaw - anchor_yaw);
    }

    if (flatten_to_2d_) {
      z = 0.0;
    }

    nav_msgs::msg::Odometry odom;
    odom.header.stamp = output_stamp(*msg);
    odom.header.frame_id = output_odom_frame_;
    odom.child_frame_id = output_base_frame_;
    odom.pose.pose.position.x = x;
    odom.pose.pose.position.y = y;
    odom.pose.pose.position.z = z;
    odom.pose.pose.orientation = quaternion_from_yaw(yaw);
    odom.pose.covariance = msg->pose.covariance;

    const auto base_rotation_from_child = base_from_child->getBasis();
    const auto base_translation_from_child = base_from_child->getOrigin();
    const tf2::Vector3 linear_child(
      msg->twist.twist.linear.x,
      msg->twist.twist.linear.y,
      msg->twist.twist.linear.z);
    const tf2::Vector3 angular_child(
      msg->twist.twist.angular.x,
      msg->twist.twist.angular.y,
      msg->twist.twist.angular.z);
    const auto linear_base = base_rotation_from_child * linear_child;
    const auto angular_base = base_rotation_from_child * angular_child;
    const auto offset_velocity = angular_base.cross(base_translation_from_child);

    odom.twist.twist.linear.x = linear_base.x() - offset_velocity.x();
    odom.twist.twist.linear.y = linear_base.y() - offset_velocity.y();
    odom.twist.twist.linear.z =
      flatten_to_2d_ ? 0.0 : linear_base.z() - offset_velocity.z();
    odom.twist.twist.angular.x = flatten_to_2d_ ? 0.0 : angular_base.x();
    odom.twist.twist.angular.y = flatten_to_2d_ ? 0.0 : angular_base.y();
    odom.twist.twist.angular.z = angular_base.z();
    odom.twist.covariance = msg->twist.covariance;

    odom_pub_->publish(odom);

    if (!tf_pub_) {
      return;
    }
    geometry_msgs::msg::TransformStamped tf;
    tf.header = odom.header;
    tf.child_frame_id = output_base_frame_;
    tf.transform.translation.x = x;
    tf.transform.translation.y = y;
    tf.transform.translation.z = z;
    tf.transform.rotation = odom.pose.pose.orientation;
    tf2_msgs::msg::TFMessage tf_msg;
    tf_msg.transforms.push_back(tf);
    tf_pub_->publish(tf_msg);
  }

  std::string input_topic_;
  std::string output_topic_;
  std::string tf_topic_;
  std::string output_odom_frame_;
  std::string output_base_frame_;
  std::string sensor_frame_;
  bool anchor_on_first_sample_{true};
  bool flatten_to_2d_{true};
  bool publish_tf_{false};
  bool restamp_output_to_now_{false};
  double output_stamp_offset_sec_{0.0};
  bool input_reliable_{false};
  bool output_reliable_{true};
  int input_qos_depth_{1};
  int output_qos_depth_{20};
  int tf_qos_depth_{100};
  double static_lookup_timeout_sec_{0.05};

  std::unique_ptr<tf2_ros::Buffer> tf_buffer_;
  std::unique_ptr<tf2_ros::TransformListener> tf_listener_;
  rclcpp::Publisher<nav_msgs::msg::Odometry>::SharedPtr odom_pub_;
  rclcpp::Publisher<tf2_msgs::msg::TFMessage>::SharedPtr tf_pub_;
  rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr odom_sub_;
  std::optional<tf2::Transform> base_from_child_;
  std::optional<std::tuple<double, double, double>> anchor_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<FastlioOdomBridge>());
  rclcpp::shutdown();
  return 0;
}

File: src/robot_fastlio_mapping/scripts/fastlio_odom_bridge_node.py
#!/usr/bin/env python3
from __future__ import annotations

import math
from typing import Optional, Tuple

import rclpy
from geometry_msgs.msg import Quaternion, TransformStamped
from nav_msgs.msg import Odometry
from rclpy.duration import Duration
from rclpy.node import Node
from rclpy.time import Time
from tf2_msgs.msg import TFMessage
from tf2_ros import Buffer, TransformException, TransformListener


Vector3 = Tuple[float, float, float]
Matrix3 = Tuple[Tuple[float, float, float], Tuple[float, float, float], Tuple[float, float, float]]
Transform = Tuple[Matrix3, Vector3]


def normalize_angle(value: float) -> float:
    return math.atan2(math.sin(value), math.cos(value))


def quaternion_to_matrix(q: Quaternion) -> Matrix3:
    x, y, z, w = q.x, q.y, q.z, q.w
    norm = math.sqrt(x * x + y * y + z * z + w * w)
    if norm <= 1.0e-12:
        return ((1.0, 0.0, 0.0), (0.0, 1.0, 0.0), (0.0, 0.0, 1.0))
    x, y, z, w = x / norm, y / norm, z / norm, w / norm
    xx, yy, zz = x * x, y * y, z * z
    xy, xz, yz = x * y, x * z, y * z
    wx, wy, wz = w * x, w * y, w * z
    return (
        (1.0 - 2.0 * (yy + zz), 2.0 * (xy - wz), 2.0 * (xz + wy)),
        (2.0 * (xy + wz), 1.0 - 2.0 * (xx + zz), 2.0 * (yz - wx)),
        (2.0 * (xz - wy), 2.0 * (yz + wx), 1.0 - 2.0 * (xx + yy)),
    )


def yaw_to_quaternion(yaw: float) -> Quaternion:
    q = Quaternion()
    half = 0.5 * yaw
    q.z = math.sin(half)
    q.w = math.cos(half)
    return q


def transform_from_pose(position, orientation: Quaternion) -> Transform:
    return quaternion_to_matrix(orientation), (float(position.x), float(position.y), float(position.z))


def transform_from_msg(msg: TransformStamped) -> Transform:
    t = msg.transform.translation
    return quaternion_to_matrix(msg.transform.rotation), (float(t.x), float(t.y), float(t.z))


def rotate(matrix: Matrix3, vector: Vector3) -> Vector3:
    return (
        matrix[0][0] * vector[0] + matrix[0][1] * vector[1] + matrix[0][2] * vector[2],
        matrix[1][0] * vector[0] + matrix[1][1] * vector[1] + matrix[1][2] * vector[2],
        matrix[2][0] * vector[0] + matrix[2][1] * vector[1] + matrix[2][2] * vector[2],
    )


def cross(a: Vector3, b: Vector3) -> Vector3:
    return (
        a[1] * b[2] - a[2] * b[1],
        a[2] * b[0] - a[0] * b[2],
        a[0] * b[1] - a[1] * b[0],
    )


def matmul(a: Matrix3, b: Matrix3) -> Matrix3:
    return tuple(
        tuple(sum(a[row][k] * b[k][col] for k in range(3)) for col in range(3)) for row in range(3)
    )  # type: ignore[return-value]


def transpose(a: Matrix3) -> Matrix3:
    return tuple(tuple(a[col][row] for col in range(3)) for row in range(3))  # type: ignore[return-value]


def compose(a: Transform, b: Transform) -> Transform:
    ar, at = a
    br, bt = b
    r = matmul(ar, br)
    rotated_bt = rotate(ar, bt)
    t = (at[0] + rotated_bt[0], at[1] + rotated_bt[1], at[2] + rotated_bt[2])
    return r, t


def inverse(t: Transform) -> Transform:
    r, v = t
    rt = transpose(r)
    inv_v = rotate(rt, (-v[0], -v[1], -v[2]))
    return rt, inv_v


def yaw_from_matrix(r: Matrix3) -> float:
    return math.atan2(r[1][0], r[0][0])


class FastlioOdomBridge(Node):
    def __init__(self) -> None:
        super().__init__("fastlio_odom_bridge")
        self.input_topic = self.declare_parameter("input_topic", "/Odometry").value
        self.output_topic = self.declare_parameter("output_topic", "/fastlio/base_odometry").value
        self.tf_topic = self.declare_parameter("tf_topic", "/tf_slam2d").value
        self.output_odom_frame = self.declare_parameter("output_odom_frame", "odom").value
        self.output_base_frame = self.declare_parameter("output_base_frame", "base_link").value
        self.sensor_frame = self.declare_parameter("sensor_frame", "lidar_link").value
        self.anchor_on_first_sample = bool(self.declare_parameter("anchor_on_first_sample", True).value)
        self.flatten_to_2d = bool(self.declare_parameter("flatten_to_2d", True).value)
        self.publish_tf = bool(self.declare_parameter("publish_tf", False).value)
        self.restamp_output_to_now = bool(self.declare_parameter("restamp_output_to_now", False).value)
        self.output_stamp_offset_sec = float(self.declare_parameter("output_stamp_offset_sec", 0.0).value)

        self.tf_buffer = Buffer(cache_time=Duration(seconds=30.0))
        self.tf_listener = TransformListener(self.tf_buffer, self)
        self.odom_pub = self.create_publisher(Odometry, str(self.output_topic), 20)
        self.tf_pub = self.create_publisher(TFMessage, str(self.tf_topic), 100) if self.publish_tf else None
        self.create_subscription(Odometry, str(self.input_topic), self.on_odom, 20)

        self.base_from_child: Optional[Transform] = None
        self.anchor: Optional[Tuple[float, float, float]] = None
        self.get_logger().info(
            "FAST-LIO odom bridge: "
            f"{self.input_topic} -> {self.output_topic}, "
            f"frame={self.output_odom_frame}->{self.output_base_frame}, publish_tf={self.publish_tf}, "
            f"restamp_output_to_now={self.restamp_output_to_now}"
        )

    def lookup_base_from_child(self, child_frame: str) -> Optional[Transform]:
        if child_frame == self.output_base_frame:
            identity: Matrix3 = ((1.0, 0.0, 0.0), (0.0, 1.0, 0.0), (0.0, 0.0, 1.0))
            return identity, (0.0, 0.0, 0.0)
        if self.base_from_child is not None:
            return self.base_from_child
        try:
            tf = self.tf_buffer.lookup_transform(
                str(self.output_base_frame), child_frame, Time(), timeout=Duration(seconds=0.05)
            )
        except TransformException as exc:
            self.get_logger().warn(
                f"waiting for static {self.output_base_frame}->{child_frame}: {exc}",
                throttle_duration_sec=2.0,
            )
            return None
        self.base_from_child = transform_from_msg(tf)
        self.get_logger().info(f"using static {self.output_base_frame}->{child_frame} to convert FAST-LIO odom")
        return self.base_from_child

    def output_stamp(self, msg: Odometry):
        if not self.restamp_output_to_now:
            return msg.header.stamp
        stamp = self.get_clock().now()
        if abs(self.output_stamp_offset_sec) > 1.0e-9:
            stamp = stamp + Duration(seconds=self.output_stamp_offset_sec)
        return stamp.to_msg()

    def on_odom(self, msg: Odometry) -> None:
        child_frame = (msg.child_frame_id or str(self.sensor_frame)).lstrip("/")
        base_from_child = self.lookup_base_from_child(child_frame)
        if base_from_child is None:
            return

        odom_from_child = transform_from_pose(msg.pose.pose.position, msg.pose.pose.orientation)
        odom_from_base = compose(odom_from_child, inverse(base_from_child))
        base_rotation_from_child, base_translation_from_child = base_from_child
        rotation, translation = odom_from_base
        x, y, z = translation
        yaw = yaw_from_matrix(rotation)

        if self.anchor_on_first_sample:
            if self.anchor is None:
                self.anchor = (x, y, yaw)
            anchor_x, anchor_y, anchor_yaw = self.anchor
            dx, dy = x - anchor_x, y - anchor_y
            c, s = math.cos(anchor_yaw), math.sin(anchor_yaw)
            x = c * dx + s * dy
            y = -s * dx + c * dy
            yaw = normalize_angle(yaw - anchor_yaw)

        if self.flatten_to_2d:
            z = 0.0

        stamp = self.output_stamp(msg)
        odom = Odometry()
        odom.header.stamp = stamp
        odom.header.frame_id = str(self.output_odom_frame)
        odom.child_frame_id = str(self.output_base_frame)
        odom.pose.pose.position.x = x
        odom.pose.pose.position.y = y
        odom.pose.pose.position.z = z
        odom.pose.pose.orientation = yaw_to_quaternion(yaw)
        odom.pose.covariance = msg.pose.covariance
        linear_child = msg.twist.twist.linear
        angular_child = msg.twist.twist.angular
        linear_base = rotate(base_rotation_from_child, (linear_child.x, linear_child.y, linear_child.z))
        angular_base = rotate(base_rotation_from_child, (angular_child.x, angular_child.y, angular_child.z))
        offset_velocity = cross(angular_base, base_translation_from_child)
        odom.twist.twist.linear.x = linear_base[0] - offset_velocity[0]
        odom.twist.twist.linear.y = linear_base[1] - offset_velocity[1]
        odom.twist.twist.linear.z = 0.0 if self.flatten_to_2d else linear_base[2] - offset_velocity[2]
        odom.twist.twist.angular.x = 0.0 if self.flatten_to_2d else angular_base[0]
        odom.twist.twist.angular.y = 0.0 if self.flatten_to_2d else angular_base[1]
        odom.twist.twist.angular.z = angular_base[2]
        odom.twist.covariance = msg.twist.covariance
        self.odom_pub.publish(odom)

        if self.tf_pub is None:
            return
        tf = TransformStamped()
        tf.header.stamp = stamp
        tf.header.frame_id = str(self.output_odom_frame)
        tf.child_frame_id = str(self.output_base_frame)
        tf.transform.translation.x = x
        tf.transform.translation.y = y
        tf.transform.translation.z = z
        tf.transform.rotation = odom.pose.pose.orientation
        self.tf_pub.publish(TFMessage(transforms=[tf]))


def main() -> None:
    rclpy.init()
    node = FastlioOdomBridge()
    try:
        rclpy.spin(node)
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()
